File: lib/java/src/main/java/io/github/msgpact/internal/binary/UnpackList.java
package io.github.msgpact.internal.binary;

import static io.github.msgpact.internal.binary.PackList.PACKED_BYTE;
import static io.github.msgpact.internal.binary.Unpack.unpack;
import static io.github.msgpact.internal.binary.UnpackMap.unpackMap;

import java.util.ArrayList;
import java.util.List;

import org.msgpack.jackson.dataformat.MessagePackExtensionType;

public class UnpackList {
    static List<Object> unpackList(List<Object> list) {
        if (list.isEmpty()) {
            return list;
        }

        if (!(list.get(0) instanceof final MessagePackExtensionType t && t.getType() == PACKED_BYTE)) {
            final var newList = new ArrayList<Object>();
            for (final var e : list) {
                newList.add(unpack(e));
            }
            return newList;
        }

        final var unpackedList = new ArrayList<Object>();
        final var headers = (List<Object>) list.get(1);

        for (int i = 2; i < list.size(); i += 1) {
            final var row = (List<Object>) list.get(i);
            final var m = unpackMap(row, headers);

            unpackedList.add(m);
        }

        return unpackedList;
    }
}

File: lib/py/telepact/internal/binary/UnpackList.py
#|
#|  Copyright The Telepact Authors
#|  SPDX-License-Identifier: Apache-2.0
#|

from typing import cast
from msgpack import ExtType
from threading import Lock

from ...internal.binary.PackList import PACKED_BYTE
from ...internal.binary.UnpackMap import unpack_map
_UNPACK = None
_UNPACK_LOCK = Lock()


def _get_unpack():
    global _UNPACK
    if _UNPACK is None:
        with _UNPACK_LOCK:
            if _UNPACK is None:
                from ...internal.binary.Unpack import unpack as _unpack
                _UNPACK = _unpack
    return _UNPACK


def unpack_list(lst: list[object]) -> list[object]:
    unpack = _get_unpack()

    if not lst:
        return lst

    first_item = lst[0]
    if type(first_item) is not ExtType or first_item.code != PACKED_BYTE:
        new_lst = []
        for item in lst:
            new_lst.append(unpack(item))
        return new_lst

    unpacked_lst: list[object] = []
    headers = cast(list[object], lst[1])

    for i in range(2, len(lst)):
        row = cast(list[object], lst[i])
        m = unpack_map(row, headers)

        unpacked_lst.append(m)

    return unpacked_lst
